Feature #2882 seeps qa

JohnHalleyGotway opened the pull request:
Note that updated SEEPS climo data can be found in `PPT24_seepsweights_grid_v12.0.nc` from https://dtcenter.ucar.edu/dfiles/code/METplus/MET/MET_unit_test/unit_test/climatology_data/seeps/PPT24_seepsweights_grid_v12.0.nc

Question: Should I REPLACE the existing `PPT24_seepsweights_grid.nc` file with the updated version in `PPT24_seepsweights_grid_v12.0.nc`? Once this PR is approved, I could make that change and update [this line](https://github.com/dtcenter/MET/blob/0d29a871050191fcc6edc223ea7f69c0161eb685/internal/test_unit/xml/unit_grid_stat.xml#L299) before merging.

## Expected Differences ##

- [x] Do these changes introduce new tools, command line arguments, or configuration file options? **[No]**</br>
If **yes**, please describe:</br>

- [x] Do these changes modify the structure of existing or add new output data types (e.g. statistic line types or NetCDF variables)? **[Yes]**</br>
If **yes**, please describe:</br>

As requested by @mpm-meto and @RachelNorth, modifies the SEEPS line type by renaming columns from `S12 S13 S21 S23 S31 S32` to `ODFL ODFH OLFD OLFH OHFD OHFL` where `O` and `F` indicate observation and forecast and `D`, `L`, and `H` indicate dry, light, and heavy precipitation. 

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>
Ran the unit tests and confirmed they all run without error.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

Please find this branch compiled on seneca in `/d1/projects/MET/MET_pull_requests/met-12.0.0/beta6/MET-feature_2882_seeps_qa`.

@j-opatz, please review the code and documentation changes for completeness. Honestly, these changes warrant further validation and testing by @mpm-meto and @RachelNorth which they'll do when testing the beta6 release. Please coordinate with the MetOffice on these details.

@hsoh-u, please review the code changes, in particular changing from dynamically allocated memory to STL vector.

- [x] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[Yes]**

Update the SEEPS column names in the user's guide.

- [x] Do these changes include sufficient testing updates? **[Yes]**

I added no new tests because no significant new functionality was added.

- [x] Will this PR result in changes to the MET test suite? **[Yes]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

The following differences will be flagged in the GHA testing workflow:
1. Grid-Stat
  - `grid_stat/grid_stat_SEEPS_000000L_20211202_000000V_seeps.txt`, `grid_stat/grid_stat_SEEPS_000000L_20211202_000000V_seeps.stat`, and `grid_stat/grid_stat_SEEPS_000000L_20211202_000000V_pairs.nc`
  - SEEPS column names are changed.
  - All of those modified columns are now populated with values of 0. @mpm-meto and @RachelNorth should advise during their testing of beta6 whether this is a desired outcome.
  - Relatively small changes in the `SEEPS_MPR_SCORE_precipitation_amount_all_all_FULL` variable.
2. Stat-Analysis output from Point-Stat input
  - `stat_analysis_ps/POINT_STAT_SEEPS.out`
  - Column names are modified, but not changes to the data.

Note that updated SEEPS climo data can be found in `PPT24_seepsweights_grid_v12.0.nc` from https://dtcenter.ucar.edu/dfiles/code/METplus/MET/MET_unit_test/unit_test/climatology_data/seeps/PPT24_seepsweights_grid_v12.0.nc

I assume the modified SEEPS output is due to the modified gridded SEEPS dataset.

- [x] Will this PR result in changes to existing METplus Use Cases? **[Maybe]**</br>
If **yes**, create a new **Update Truth** [METplus issue](https://github.com/dtcenter/METplus/issues/new/choose) to describe them.

If a METplus use case exists for SEEPS, then it's output will change. I'm not sure that it actually does though.

- [x] Do these changes introduce new SonarQube findings? **[No]**</br>
If **yes**, please describe:

SonarQube originally flagged 25 "new" code smells. I did another pass through to fix the easy ones and reduce that to 15, and I note that the this PR reduces the overall number of code smells from 18,997 (in the develop branch) to 18,909 (in this feature branch). The reduction is due primarily to switching from dynamically allocated memory to STL vectors.

- [x] Please complete this pull request review by **[Fri 10/11/24]**.</br>

## Pull Request Checklist ##
See the [METplus Workflow](https://metplus.readthedocs.io/en/latest/Contributors_Guide/github_workflow.html) for details.
- [x] Review the source issue metadata (required labels, projects, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)** and **Development** issue
Select: **Milestone** as the version that will include these changes
Select: **Coordinated METplus-X.Y Support** project for bugfix releases or **MET-X.Y.Z Development** project for official releases
- [x] After submitting the PR, select the :gear: icon in the **Development** section of the right hand sidebar. Search for the issue that this PR will close and select it, if it is not already selected.
- [x] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [x] Close the linked issue and delete your feature or bugfix branch from GitHub.

mpm-meto committed:
Update seeps.h

Change variable names to reduce ambiguity for interpretation and aid useability.

mpm-meto committed:
Update seeps.cc

Pull through variable name changes and renaming of functions to aid legibility and clarity. Introduced some additional debug print statements.

mpm-meto committed:
Update grid-stat.rst

Add documentation about the location of the gridded climatology files for SEEPS and which environment variable to use.

mpm-meto committed:
Replace read_seeps_scores() with get_seeps_climo_grid()

JohnHalleyGotway committed:
Merge remote-tracking branch 'origin/patch-2' into merge-patch-1-4

JohnHalleyGotway committed:
Manually merging Rachel's patch-1 changes.

JohnHalleyGotway committed:
Getting close to getting these seeps changes to compile. But it's failing in pair_data_point.cc

JohnHalleyGotway committed:
Merge remote-tracking branch 'origin/develop' into feature_2882_seeps_qa

JohnHalleyGotway committed:
Per #2882, get branch feature_2882_seeps_qa compiling again. Recommend revisiting the volume of SEEPS-related Debug log messages and reducing them once its fully tested.

JohnHalleyGotway committed:
Per #2882, need to update the handling of the PPT24_seepsweights_grid.nc file name. Rename as _v12.0.nc for the updated version with the new names so that the existing regressions tests and nightly builds for main_v11.1 and develop continue to work. We can remove the _v12.0 once this feature branch is merged into develop but for the time being, we need both versions to exist.

JohnHalleyGotway committed:
Merge remote-tracking branch 'origin/develop' into feature_2882_seeps_qa

JohnHalleyGotway committed:
Per #2882, rename the SEEPS columns from S12, S13, S21, S23, S31, S32 to the more descriptive ODFL, ODFH, OLFD, OLFH, OHFD, OHFL names.

JohnHalleyGotway committed:
Per #2882, update SEEPS details

JohnHalleyGotway committed:
Per #2882, store and report the weighted mean fcst and mean obs, just like the SEEPS score itself so that they're handled in a consistent manner. Note however that it's hard-coded to NOT write the weighted means/score, only the unweighted ones.

JohnHalleyGotway committed:
Per #2882, change SEEPS debug log levels and correct the storage of mean_fcst and mean_obs values.

JohnHalleyGotway committed:
Per #2882, correct SEEPS column name lookups

JohnHalleyGotway committed:
Per #2882, call is_bad_data() instead of is_eq(..., -9999.0) to get rid of compiler warning message.

JohnHalleyGotway committed:
Per #2882, add 2 more variations of the is_eq() function with mixed float and double inputs to satisfy compiler pb2nc compiler warnings.

JohnHalleyGotway committed:
Merge remote-tracking branch 'origin/develop' into feature_2882_seeps_qa

JohnHalleyGotway committed:
Merge remote-tracking branch 'origin/develop' into feature_2882_seeps_qa

JohnHalleyGotway committed:
Per #2882, switch from dynamically allocated arrays to std::vector

JohnHalleyGotway committed:
Per #2882, enhance Stat-Analysis to write the SEEPS line type to an output .stat file.

JohnHalleyGotway committed:
Per #2882, update the aggregated seeps computation to use better-initialized vectors.

hsoh-u reviewed (dismissed):
I approve the changes.
- the variable names are more meaningful
- Reduced SonarQube findings with different dynamic memory allocation method (using vector)
- Include bug fixes

JohnHalleyGotway committed:
Per #2882, resolve a few more SonarQube code smells.

j-opatz reviewed (dismissed):
Documentation looks good, updates to the column names reflects the issue changes requested.

As to the question posed, I think swapping out the data as soon as this PR is approved is the best route. Using new data that reflects what a user will encounter when looking for data will help the METplus experience be more seamless.

JohnHalleyGotway committed:
Per #2882, now that this PR is ready to merge, remove the v12.0 version number from the gridded SEEPS climo file name ci-skip-all

JohnHalleyGotway commented:
Note that I renamed the SEEPS climo file to remove the specific version (`v12.0`) from the filename, and updated the code in this branch accordingly:
`PPT24_seepsweights_grid_v12.0.nc` > `PPT24_seepsweights_grid.nc`

The existing file is renamed as:
`PPT24_seepsweights_grid.nc` > `PPT24_seepsweights_grid_pre_v12.0.nc`

Proceeding with merging this PR now.
